File: src/shared/services/llm/adapters/qwen-adapter.ts
/**
 * 阿里云通义千问适配器
 */

import { BaseLLMAdapter } from '../base-adapter';
import type { LLMRequest, LLMResponse } from '../types';

export class QwenAdapter extends BaseLLMAdapter {
  private baseUrl: string;

  constructor(config: any) {
    super(config);
    this.baseUrl = config.baseUrl || 'https://dashscope.aliyuncs.com/api/v1';
  }

  async complete(request: LLMRequest): Promise<LLMResponse> {
    try {
      await this.validateConfig();

      return await this.retry(async () => {
        return await this.withTimeout(this._sendRequest(request));
      });
    } catch (error) {
      this.handleError(error, '通义千问API调用失败');
    }
  }

  private async _sendRequest(request: LLMRequest): Promise<LLMResponse> {
    const headers: Record<string, string> = {
      'Authorization': `Bearer ${this.config.apiKey}`,
      'X-DashScope-SSE': 'disable',
    };
    if (this.config.customHeaders) Object.assign(headers, this.config.customHeaders);

    const response = await fetch(`${this.baseUrl}/services/aigc/text-generation/generation`, {
      method: 'POST',
      headers: this.buildHeaders(headers),
      body: JSON.stringify({
        model: this.config.model,
        input: {
          messages: request.messages.map(msg => ({
            role: msg.role,
            content: msg.content,
          })),
        },
        parameters: {
          temperature: request.temperature ?? this.config.temperature,
          max_tokens: request.maxTokens ?? this.config.maxTokens,
          top_p: request.topP ?? this.config.topP,
          result_format: 'message',
        },
      }),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw {
        statusCode: response.status,
        message: error.message || `HTTP ${response.status}: ${response.statusText}`,
      };
    }

    const data = await response.json();

    if (data.code && data.code !== '200') {
      throw new Error(`API错误 (${data.code}): ${data.message}`);
    }

    const output = data.output;
    if (!output?.choices?.[0]) {
      throw new Error('API响应格式异常: 缺少output.choices字段');
    }

    const choice = output.choices[0];

    return {
      content: choice.message?.content || '',
      model: this.config.model,
      usage: output.usage ? {
        promptTokens: output.usage.input_tokens,
        completionTokens: output.usage.output_tokens,
        totalTokens: output.usage.total_tokens || 
          (output.usage.input_tokens + output.usage.output_tokens),
      } : undefined,
      finishReason: choice.finish_reason,
    };
  }

  async validateConfig(): Promise<boolean> {
    await super.validateConfig();
    
    if (!this.config.model) {
      throw new Error('未指定通义千问模型');
    }
    
    return true;
  }
}

File: backend/app/services/llm/adapters/qwen_adapter.py
"""
阿里云通义千问适配器 - 兼容OpenAI格式
"""

from typing import Dict, Any
from ..base_adapter import BaseLLMAdapter
from ..types import LLMRequest, LLMResponse, LLMUsage, DEFAULT_BASE_URLS, LLMProvider


class QwenAdapter(BaseLLMAdapter):
    """通义千问适配器"""
    
    @property
    def base_url(self) -> str:
        return self.config.base_url or DEFAULT_BASE_URLS.get(LLMProvider.QWEN, "https://dashscope.aliyuncs.com/compatible-mode/v1")
    
    async def complete(self, request: LLMRequest) -> LLMResponse:
        try:
            await self.validate_config()
            return await self.retry(lambda: self._send_request(request))
        except Exception as error:
            self.handle_error(error, "通义千问 API调用失败")
    
    async def _send_request(self, request: LLMRequest) -> LLMResponse:
        # 通义千问兼容OpenAI格式
        headers = {
            "Authorization": f"Bearer {self.config.api_key}",
        }
        
        messages = [{"role": msg.role, "content": msg.content} for msg in request.messages]
        
        request_body: Dict[str, Any] = {
            "model": self.config.model,
            "messages": messages,
            "temperature": request.temperature if request.temperature is not None else self.config.temperature,
            "max_tokens": request.max_tokens if request.max_tokens is not None else self.config.max_tokens,
            "top_p": request.top_p if request.top_p is not None else self.config.top_p,
        }
        
        url = f"{self.base_url.rstrip('/')}/chat/completions"
        
        response = await self.client.post(
            url,
            headers=self.build_headers(headers),
            json=request_body
        )
        
        if response.status_code != 200:
            error_data = response.json() if response.text else {}
            error_msg = error_data.get("error", {}).get("message", f"HTTP {response.status_code}")
            raise Exception(f"{error_msg}")
        
        data = response.json()
        choice = data.get("choices", [{}])[0]
        
        if not choice:
            raise Exception("API响应格式异常: 缺少choices字段")
        
        usage = None
        if "usage" in data:
            usage = LLMUsage(
                prompt_tokens=data["usage"].get("prompt_tokens", 0),
                completion_tokens=data["usage"].get("completion_tokens", 0),
                total_tokens=data["usage"].get("total_tokens", 0)
            )
        
        return LLMResponse(
            content=choice.get("message", {}).get("content", ""),
            model=data.get("model"),
            usage=usage,
            finish_reason=choice.get("finish_reason")
        )
    
    async def validate_config(self) -> bool:
        await super().validate_config()
        if not self.config.model:
            raise Exception("未指定通义千问模型")
        return True


